chore: switch to ruff for code formatting

wyardley opened the pull request:
- Rename `flake8` to `ruff-check` since that's what it was already running
- Add new `ruff-format` and `ruff-format-check` make targets
- Update lint target names
- Update README badge
- Use ruff for test files as well (minimal changes will be needed)
- Update ruff to `v0.9.9`

I'll push the formatting changes in a separate commit. However, there are now some `mypy` failures that I may need some help addressing or pinning something for.

wyardley commented on pyproject.toml:
```diff
@@ -48,16 +48,16 @@ importlib-metadata = "<8.6"
 freezegun = "*"
 pytest = "*"
 pylint = "^3.3.3"
+ruff = "==0.9.6"
```
good to pin exactly, and match the pre-commit version here.

jamesoff commented:
Looks like the problem is ring-doorbell > 0.6.0 which changed API; I don't have a Ring doorbell myself now (nor even an account) so for now I've disabled the monitor entirely (it wouldn't work anyway I don't think). Dependency resolution means that having ring-doorbell < 0.6.0 causes other issues, which is why I went that route.

If you pull in that commit (cb733a0) or rebase on develop that should "fix" the mypy issues :)

jamesoff commented:
I see PyArlo is broken on 3.9 too actually, thanks to using the new typing syntax. Feel free to bump the required Python 3.10 here or separately (not sure what will work best for you alongside this PR) since 3.9 is definitely getting on a bit, and I see the 3.10 tests were going to pass before Github interrupted them because of the 3.9 failure.

wyardley commented:
I rebased, and I updated ruff to 0.9.9 and updated the formatting again for one minor change.
I'm thinking if the plan is to drop Python 3.9, it should be a separate commit marked as a breaking change, so maybe you want to do that separately and then I can rebase again?

Side note: probably should have a `requires-python` line in `pyproject.toml`? I see the
```
[tool.poetry.dependencies]
python = "^3.9.0"
```
line but don't think there's an explicit requirement of Python versions?

So probably something like this would work (assuming as a breaking change?)

```diff
diff --git a/.github/workflows/linting.yml b/.github/workflows/linting.yml
index 294985d..3f5220f 100644
--- a/.github/workflows/linting.yml
+++ b/.github/workflows/linting.yml
@@ -8,7 +8,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ['3.9', '3.10', '3.11', '3.12', '3.13']
+        python-version: ['3.10', '3.11', '3.12', '3.13']
     name: Python ${{ matrix.python-version }}
 
     steps:
diff --git a/.github/workflows/main.yml b/.github/workflows/main.yml
index a8a589c..a44b47c 100644
--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -8,7 +8,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ['3.9', '3.10', '3.11', '3.12', '3.13']
+        python-version: ['3.10', '3.11', '3.12', '3.13']
     name: Python ${{ matrix.python-version }}
 
     steps:
diff --git a/.github/workflows/windows.yml b/.github/workflows/windows.yml
index a904bf6..6bb768f 100644
--- a/.github/workflows/windows.yml
+++ b/.github/workflows/windows.yml
@@ -8,7 +8,7 @@ jobs:
     runs-on: windows-latest
     strategy:
       matrix:
-        python-version: ['3.9', '3.10', '3.11', '3.12', '3.13']
+        python-version: ['3.10', '3.11', '3.12', '3.13']
     name: Python ${{ matrix.python-version }} (Windows)
 
     steps:
diff --git a/pyproject.toml b/pyproject.toml
index 066930e..88d01fb 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -4,6 +4,7 @@ version = "1.14.0"
 description = "A simple network and host monitor"
 license = "BSD-3-Clause"
 authors = ["James Seward <james@jamesoff.net>"]
+requires-python = ">=3.10"
 readme = "README.md"
 homepage = "https://github.com/jamesoff/simplemonitor"
 repository = "https://github.com/jamesoff/simplemonitor"
@@ -30,7 +31,7 @@ simplemonitor = "simplemonitor.monitor:main"
 winmonitor = "simplemonitor.winmonitor:main"
 
 [tool.poetry.dependencies]
-python = "^3.9.0"
+python = "^3.10.0"
 arrow = ">=0.17,<1.4"
 boto3 = "^1.15.16"
 colorlog = ">=4.4"
```

wyardley commented:
ack - seems like poetry is also using the older, deprecated syntax? I have 1.8.5 locally too, but I think we should move most or all the stuff in `[tool.poetry]` to `[project]` in `pyproject.toml`? 

In [1.x, most of those fields are duplicated](https://python-poetry.org/docs/1.8/pyproject/), but in 2.1.x https://python-poetry.org/docs/pyproject/

I'm not seeing where the poetry version in the CI pipelines is coming from, but either we would need to add the `[project]` section (with most of the same fields / values), or update / require poetry 2.x.

wyardley commented:
@jamesoff what do you think of ce0bc6d8d948655d876a55d260f0776db8738414? I can revert if if you'd rather do this yourself. Also let me know if you want it to be type feat / chore instead (since it's a breaking change, it would be triggering either way), or if you don't want it to be considered a breaking change. I used both the bang syntax and `BREAKING CHANGE:` in the body, so if you're using any implementation of semantic-release, I think we should be good.

For this to work, you'll probably need to do a merge commit or rebase merge vs. squash merging these 3 commits.

That commit basically
* Drops support for Python 3.9
* Enforces Poetry 2.x for development
* Migrates to the new style Poetry config
* Adds `require-python` to the project settings explicitly

Side note: I'm guessing some / all of what's in `setup.cfg` might be able to be migrated to `pyproject.toml` too? and maybe the flake8 stuff in there could be dropped? Not sure if you're using `isort`, but we could also switch to configuring ruff to handle any isort / flake8 rules you want enforced, I think, if it's not already.

I could do some of that cleanup in a followup, since there's already a lot going on here.

wyardley commented on simplemonitor/Loggers/file.py:
```diff
@@ -103,9 +101,7 @@
-            self.file_handle = open(
-                self.filename, "a+"
-            )  # pylint: disable=consider-using-with
+            self.file_handle = open(self.filename, "a+")  # pylint: disable=consider-using-with
```
Seems like this behavior already existed.
If not using with, not sure if it might be good to add an explicit close somewhere?

jamesoff replied:
I had a feeling I looked at this previously and decided I was happy with the flow in the code - that's why the pylint comment is there - but may be worth revisiting it.

jamesoff commented:
That commit looks good to me, I'm happy with how you have it (I've not gotten myself disciplined enough to use that style of commit message, although I probably should, especially as I appreciate it in other projects :)

I'm using isort only because it's there, can ruff do import sorting now? (Mostly I just wanted to have imports organised/sorted in a sensible way, I'm not wedded to isort specifically.)

Modernising(?) by pulling setup.cfg apart into other files is good too - by way of background simplemonitor was the first real Python program I wrote, and I wrote it as much to teach myself Python as fulfill a need for monitoring, so it's been through a bunch of rewrites and styles as I figured out how to do things better (and as Python changed; when I started I think it was on 2.3...) as well as working out how to package it etc.

I just cut a release of the current state, partly since I know @cpina was hoping for one and partly because if we're going to bump the minimum Python version then at least that's a recent release which supports 3.9, so I'm good to merge this now if you're happy with that, and we can iterate from there.

Appreciate all the work and commentary :)

wyardley commented:
Yeah, among other things, ruff can do it.

https://stackoverflow.com/questions/78156811/how-do-i-isort-using-ruff

IIRC, can add "I"or "I001" to `tools.ruff.lint`, e.g.,
```
[tools.ruff.lint]
select = ["D", "E", "I001"]
```
Not 100% sure if all the stuff configured curently in `setup.cfg` => `[isort]` can be configured for ruff, but I'm assuming it probably can.
While the full subset of pylint stuff is still not totally implemented, there are a lot of different rule categories not enabled by default:
https://docs.astral.sh/ruff/rules

wyardley commented:
> so I'm good to merge this now if you're happy with that, and we can iterate from there.

Works for me! Thanks.
